Funcky/DownCast.cs

```diff
@@ -0,0 +1,36 @@
+namespace Funcky;
+
+public static class DownCast<TResult>
+    where TResult : class
+{
+    public static Option<TResult> From<TItem>(Option<TItem> option)
+        where TItem : class
+        => option.SelectMany(OptionDownCast);
+
+    public static Result<TResult> From<TItem>(Result<TItem> result)
+        where TItem : class
+        => result.SelectMany(ResultDownCast);
+
+    public static Either<TLeft, TResult> From<TLeft, TRight>(Either<TLeft, TRight> either, Func<TLeft> failedCast)
+        where TRight : class
+        where TLeft : notnull
+        => either.SelectMany(EitherDownCast<TLeft, TRight>(failedCast));
+
+    private static Option<TResult> OptionDownCast<TItem>(TItem item)
+        where TItem : class
+        => Option.FromNullable(item as TResult);
+
+    private static Result<TResult> ResultDownCast<TValidResult>(TValidResult result)
+        where TValidResult : class
+        => result as TResult is { } validResult
+            ? Result.Ok(validResult)
```

Comment:
Would the implicit conversion work here?

Reply:
it works an can be rewritten as null-coalesce this way: d83a4fd02638894a1d53c48c599b8fb960a2ce6c